src/prg_quantumdynamics_mod.F90

```diff
@@ -0,0 +1,444 @@
+    real(dp)                                    :: thresh, hbar, dt
+    complex(dp), allocatable, intent(in)        :: H(:,:), SINV(:,:), rho_old(:,:), rho_cur(:,:)
+    complex(dp), allocatable, intent(out)       :: rho_new(:,:)
+    complex(dp), allocatable                    :: partial_rho(:,:), com(:,:)
+    complex(dp)                                 :: dR
+    call prg_commutate(H,rho_cur,com)
+
+    dR = cmplx(0.0_dp,2/hbar)
+    partial_rho = -dR*dt*com
+    rho_new = rho_old + partial_rho
+
+  end subroutine prg_lvni
+
+  !> Produce an excitation in the initially calculated density matrix to simulate
+  !! photo-excitation.
+  !! This routine does:
+  !! Manually induces electronic excitation based on the initial filling matrix
+  !! by promoting an electron.
+  !! \param fill_mat the initial filling matrix
+  !! \param orbit_orig the origin orbital with an electron to be promoted
+  !! \param orbit_exci the destination orbital of the promoted electron
+  !!
+  subroutine prg_excitation(fill_mat, orbit_orig, orbit_exci)
+    implicit none
+    integer, intent(inout)      :: fill_mat(:)
+    integer, intent(in)         :: orbit_orig, orbit_exci
+
+    fill_mat(orbit_orig) = fill_mat(orbit_orig) - 1.0d0
+    fill_mat(orbit_exci) = fill_mat(orbit_exci) + 1.0d0
+
+  end subroutine prg_excitation
+
+
+  !> Provides perturbation to initial density matrix in the form of an electric
+  !! field kick given input matricies in BML format.
+  !! This routine does:
+  !! \f$\hat{\rho_{kick}} =
+  !! \exp{\frac{-i}{\hbar}\hat{V}}\hat{\rho}\hat{S}\exp{\frac{i}{\hbar}\hat{V}}\hat{S^{-1}}\f$
+  !! where \f$\hat{V}\f$ is the field disturbance.
+  !! \param kick_direc the direction of the kick in the electric field
+  !! \param kick_mag the magnitude of the kick in the electric field
+  !! \param RHO the initial density matrix to be kicked in BML format.
+  !! \param S the overlap matrix
+  !! \param SINV the inverse of the overlap matrix
+  !! \param which_atom vector containing atom identification
+  !! \param r direction vector for kick based on atom and kick_direc
+  !! \param matrix_type the type of BML format
+  !! \param thresh the threshold for the BML matrix
+  !!
+  subroutine prg_kick_density_bml(kick_direc,kick_mag,RHO,S,SINV,which_atom,&
+             r,matrix_type,thresh)
+    implicit none
+    integer                             :: i, N
+    integer, intent(in)                 :: kick_direc
+    integer, allocatable, intent(in)    :: which_atom(:)
+    real(dp), allocatable               :: r(:,:)
+    real(dp)                            :: kick_mag, thresh
+    complex(dp), allocatable            :: tmat1(:,:), tmat2(:,:)
+    complex(dp)                         :: telem
+    type(bml_matrix_t)                  :: T1, T2, RHO, S, SINV
+    character(len=*), intent(in)        :: matrix_type
+
+    N=bml_get_N(RHO)
+    allocate (tmat1(N,N))
+    allocate (tmat2(N,N))
+
+    tmat1 = 0.0_dp
+    tmat2 = 0.0_dp
+    do i=1,N
```

Comment:
Here, lets use the bml_add_diagonal function. The way we do this is by filling a vector lets say telem(i) = .... 
and then we do bml_add_diagonal(telem, tmat1_bml). We should avoid allocating dense so that we don't consume memory.